kernel/src/utilities/capability_ptr.rs

```diff
@@ -45,60 +42,70 @@ pub enum CapabilityPtrPermissions {
     Execute,
 }
 
-impl From<CapabilityPtr> for usize {
-    /// Returns the address of the [`CapabilityPtr`].
-    /// Provenance note: may not expose provenance.
-    #[inline]
-    fn from(from: CapabilityPtr) -> Self {
-        from.ptr.into()
+impl Default for CapabilityPtr {
+    /// Returns a null CapabilityPtr.
+    fn default() -> Self {
+        Self { ptr: null() }
     }
 }
 
 impl From<usize> for CapabilityPtr {
-    /// Constructs a [`CapabilityPtr`] with a given address and no authority
-    ///
-    /// Provenance note: may have null provenance.
+    /// Constructs a [`CapabilityPtr`] with a given address but no authority or
+    /// provenance.
     #[inline]
     fn from(from: usize) -> Self {
-        Self { ptr: from.into() }
+        Self {
+            // Ideally this would be core::ptr::without_provenance(from), but
+            // the CHERI toolchain is too old for without_provenance. This is
+            // equivalent.
+            ptr: null::<()>().with_addr(from),
+        }
     }
 }
 
-impl From<MachineRegister> for CapabilityPtr {
-    /// Convert a register to a [`CapabilityPtr`].
-    #[inline]
-    fn from(from: MachineRegister) -> Self {
-        Self { ptr: from }
+// In addition to its publicly-documented capabilities, CapabilityPtr's
+// implementation can also store integers. MachineRegister uses that capability
+// to simplify its implementation. No other user of CapabilityPtr should rely on
+// that ability.
+
+impl From<usize> for MachineRegister {
+    fn from(from: usize) -> Self {
+        Self::from(CapabilityPtr::from(from))
     }
 }
 
 impl UpperHex for CapabilityPtr {
     #[inline]
     fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
-        UpperHex::fmt(&self.ptr, f)
+        UpperHex::fmt(&self.ptr.addr(), f)
     }
 }
 
 impl LowerHex for CapabilityPtr {
     #[inline]
     fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
-        LowerHex::fmt(&self.ptr, f)
+        LowerHex::fmt(&self.ptr.addr(), f)
     }
 }
 
 impl AddAssign<usize> for CapabilityPtr {
-    /// Increments the address of a [`CapabilityPtr`]
+    /// Increments the address of a [`CapabilityPtr`]. If the pointer is offset
+    /// past its bounds, its authority is invalidated.
```

Comment:
```suggestion
    /// past its bounds, its authority may be invalidated.
```

This is still just "possible", not guaranteed, right?

Reply:
Correct, changed.